File: Part 1/Source Code/MyApplication6/app/src/main/java/com/example/myapplication/MainActivity.java
package com.example.myapplication;

import androidx.appcompat.app.AppCompatActivity;

import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import com.example.myapplication.Helper.MyService;
import com.example.myapplication.Helper.MyServiceCallback;


public class MainActivity extends AppCompatActivity implements MyServiceCallback {


    private MyService myService;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Create an instance of the service
        myService = new MyService();

        // Set the activity as the callback listener
        myService.setCallback(this);

        // Start the service
        startService(new Intent(MainActivity.this, MyService.class));

        // Button click to trigger service button click
        Button sendService = findViewById(R.id.Btn_service);
        sendService.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // Trigger button click in service
                myService.buttonClicked();
            }
        });
    }

    // Implementation of callback method to receive data from service
    @Override
    public void onDataReceived(String data) {
        // Display data received from service
        Toast.makeText(this,  data, Toast.LENGTH_SHORT).show();
    }

}

File: Part 2/7 Mvvm/Source Code/MyApplication7/app/src/main/java/com/example/myapplication/Activity/MainActivity.kt
package com.example.myapplication.Activity

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.myapplication.Adapter.JokeAdapter
import com.example.myapplication.R
import com.example.myapplication.ViewModel.JokeViewModel

// MainActivity.kt
class MainActivity : AppCompatActivity() {

    private lateinit var viewModel: JokeViewModel
    private lateinit var adapter: JokeAdapter
    private lateinit var recyclerViewJokes: RecyclerView
    private lateinit var editTextSearch: EditText
    private lateinit var buttonSearch: Button
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        recyclerViewJokes = findViewById(R.id.recyclerViewJokes)
        editTextSearch = findViewById(R.id.editTextSearch)
        buttonSearch = findViewById(R.id.buttonSearch)

        viewModel = ViewModelProvider(this).get(JokeViewModel::class.java)
        adapter = JokeAdapter(emptyList())
        recyclerViewJokes.layoutManager = LinearLayoutManager(this)
        recyclerViewJokes.adapter = adapter

        buttonSearch.setOnClickListener {
            val query = editTextSearch.text.toString().trim()
            if (query.isNotEmpty()) {
                viewModel.searchJokes(query).observe(this, Observer { jokes ->
                    adapter.apply {
                        jokes?.let { updateJokes(it) }
                        notifyDataSetChanged()
                    }
                })
            }
        }
    }
}
